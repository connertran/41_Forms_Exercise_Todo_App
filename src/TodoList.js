import { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";
import { v4 as uuid } from "uuid";

const TodoList = () => {
  const initialState = [];
  const [todos, setTodos] = useState(initialState);

  const addTask = (text) => {
    setTodos((todos) => [...todos, { id: uuid(), text }]);
  };

  const deleteTask = (id) => {
    setTodos((todos) => todos.filter((todo) => todo.id !== id));
  };
  const editTask = (id, newText) => {
    setTodos((todos) =>
      todos.map((todo) => (todo.id === id ? { ...todo, text: newText } : todo))
    );
  };
  return (
    <>
      <NewTodoForm addTask={addTask} />
      <h2>Todo List</h2>

      {todos.map(({ id, text }) => (
        <Todo
          key={id}
          id={id}
          text={text}
          deleteTask={() => deleteTask(id)}
          editTask={editTask}
        />
      ))}
    </>
  );
};

export default TodoList;
