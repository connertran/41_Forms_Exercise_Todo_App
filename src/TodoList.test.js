import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

// smoke test
it("renders without crashing", function () {
  render(<TodoList />);
});

// snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new task", function () {
  const { getByLabelText, queryByText } = render(<TodoList />);

  // no tasks yet
  expect(queryByText("X")).not.toBeInTheDocument();

  const taskInput = getByLabelText("Task:");
  const addBtn = queryByText("Add Task!");

  // fill out the form
  fireEvent.change(taskInput, { target: { value: "homework" } });
  fireEvent.click(addBtn);

  // delete button next the task exists!
  expect(queryByText("homework")).toBeInTheDocument();
  expect(queryByText("X")).toBeInTheDocument();
});

it("can delete a task", function () {
  const { getByLabelText, queryByText } = render(<TodoList />);

  // no tasks yet
  expect(queryByText("X")).not.toBeInTheDocument();

  const taskInput = getByLabelText("Task:");
  const addBtn = queryByText("Add Task!");

  // fill out the form
  fireEvent.change(taskInput, { target: { value: "homework" } });
  fireEvent.click(addBtn);

  // delete button next the task exists!
  expect(queryByText("homework")).toBeInTheDocument();
  expect(queryByText("X")).toBeInTheDocument();

  //   delete the task
  const deleteBtn = queryByText("X");
  fireEvent.click(deleteBtn);
  //   the task is deleted
  expect(queryByText("X")).not.toBeInTheDocument();
  expect(queryByText("homework")).not.toBeInTheDocument();
});

it("can edit a task", function () {
  const { getByLabelText, queryByText } = render(<TodoList />);

  // no tasks yet
  expect(queryByText("X")).not.toBeInTheDocument();

  const taskInput = getByLabelText("Task:");
  const addBtn = queryByText("Add Task!");

  // fill out the form
  fireEvent.change(taskInput, { target: { value: "homework" } });
  fireEvent.click(addBtn);

  // delete button next the task exists!
  expect(queryByText("homework")).toBeInTheDocument();
  expect(queryByText("X")).toBeInTheDocument();

  //   edit the task
  const editBtn = queryByText("Edit");
  fireEvent.click(editBtn);
  const editInput = getByLabelText("Edit Task:");
  const saveBtn = queryByText("Save");
  fireEvent.change(editInput, { target: { value: "coding" } });
  fireEvent.click(saveBtn);
  //   the task is changed
  expect(queryByText("homework")).not.toBeInTheDocument();
  expect(queryByText("coding")).toBeInTheDocument();
});
