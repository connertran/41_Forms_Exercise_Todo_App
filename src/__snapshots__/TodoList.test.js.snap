// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="e77de432-0172-4980-96bc-33c2a960156e"
    >
      Task: 
    </label>
    <input
      id="e77de432-0172-4980-96bc-33c2a960156e"
      name="task"
      placeholder=" Enter a task to do"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Add Task!
    </button>
  </form>
  <h2>
    Todo List
  </h2>
</DocumentFragment>
`;
